drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1552 @@
+
+	return 0;
+}
+
+int sdmmc_host_set_bus_width(sdmmc_dev_t *sdio_hw, int slot, size_t width)
+{
+	if (!(slot == 0 || slot == 1)) {
+		return ESP_ERR_INVALID_ARG;
+	}
+
+	const uint16_t mask = BIT(slot);
+	uint16_t temp;
+
+	if (width == 1) {
+		temp = sdio_hw->ctype.card_width_8 & ~mask;
+		HAL_FORCE_MODIFY_U32_REG_FIELD(sdio_hw->ctype, card_width_8, temp);
+
+		temp = sdio_hw->ctype.card_width & ~mask;
+		HAL_FORCE_MODIFY_U32_REG_FIELD(sdio_hw->ctype, card_width, temp);
+
+	} else if (width == 4) {
+		temp = sdio_hw->ctype.card_width_8 & ~mask;
+		HAL_FORCE_MODIFY_U32_REG_FIELD(sdio_hw->ctype, card_width_8, temp);
+
+		temp = sdio_hw->ctype.card_width | mask;
+		HAL_FORCE_MODIFY_U32_REG_FIELD(sdio_hw->ctype, card_width, temp);
+	} else if (width == 8) {
+		temp = sdio_hw->ctype.card_width_8 | mask;
+		HAL_FORCE_MODIFY_U32_REG_FIELD(sdio_hw->ctype, card_width_8, temp);
+	} else {
+		return ESP_ERR_INVALID_ARG;
+	}
+
+	LOG_DBG("slot=%d width=%d", slot, width);
+	return ESP_OK;
+}
+
+static void configure_pin_iomux(int gpio_num)
+{
+	const int sdmmc_func = SDMMC_LL_IOMUX_FUNC;
+	const int drive_strength = 3;
+
+	if (gpio_num == GPIO_NUM_NC) {
+		return; /* parameter check*/
+	}
+
+	int rtc_num = rtc_io_num_map[gpio_num];
+
+	rtcio_hal_pulldown_disable(rtc_num);
+	rtcio_hal_pullup_enable(rtc_num);
+
+	uint32_t reg = GPIO_PIN_MUX_REG[gpio_num];
+
+	PIN_INPUT_ENABLE(reg);
+	gpio_hal_iomux_func_sel(reg, sdmmc_func);
+	PIN_SET_DRV(reg, drive_strength);
+}
+
+/**********************************************************************
+ * Zephyr API
+ **********************************************************************/
+
+/*
+ * Reset USDHC controller
+ */
+static int sdhc_esp32_reset(const struct device *dev)
+{
+	const struct sdhc_esp32_config *cfg = dev->config;
+	sdmmc_dev_t *sdio_hw = (sdmmc_dev_t *)cfg->sdio_hw;
+
+	/* Set reset bits */
+	sdio_hw->ctrl.controller_reset = 1;
+	sdio_hw->ctrl.dma_reset = 1;
+	sdio_hw->ctrl.fifo_reset = 1;
+
+	/* Wait for the reset bits to be cleared by hardware */
+	int64_t yield_delay_us = 100 * 1000; /* initially 100ms */
+	int64_t t0 = esp_timer_get_time();
+	int64_t t1 = 0;
+
+	while (sdio_hw->ctrl.controller_reset || sdio_hw->ctrl.fifo_reset ||
+	       sdio_hw->ctrl.dma_reset) {
+		t1 = esp_timer_get_time();
+
+		if (t1 - t0 > SDMMC_HOST_RESET_TIMEOUT_US) {
+			return -ETIMEDOUT;
+		}
+
+		if (t1 - t0 > yield_delay_us) {
+			yield_delay_us *= 2;
+			k_busy_wait(1);
+		}
+	}
+
+	/* Reset carried out successfully */
+	return 0;
+}
+
+/*
+ * Set SDHC io properties
+ */
+static int sdhc_esp32_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	const struct sdhc_esp32_config *cfg = dev->config;
+	sdmmc_dev_t *sdio_hw = (sdmmc_dev_t *)cfg->sdio_hw;
+	struct sdhc_esp32_data *data = dev->data;
+	uint8_t bus_width;
+	int slot = 0;
+	int ret = 0;
+
+	/* ESP32 supports 2 slots, but only one is supported at the moment.
+	 * Clock operations will affect both slots. Individual settings
```

Comment:
Interested to understand why this is- is it possible for the SDHC peripheral to have different clock configurations per each slot? Or do the the configurations need to be the same between each slot

Reply:
Both slots share the same primary clock divider. Is is possible to use different clocks but this will require a clock reconfig management each time a command is requested, if the primary divider isn't the same.